### ISSUE TITLE
BHoM: New TimeSeriesData object

### ISSUE THREAD

**Issue by rboulton-BH** [BHoM/BHoM#1189]
#### Description:

Results from many Building Environments and Specialists analysis are dynamic i.e. they vary over time.  Additionally, there is no standard time step or range so this must remain flexible. 

This functionality is not currently supported by existing objects in BHoM but is clearly needed to support multiple workflows e.g. CFD, daylight, thermal simulation etc. Uses for this also extend beyond analysis results, for example it can be used to support the handling of meter data (temperature, CO2, energy, heat etc.)

The functionality should ultimately be similar to that of a Pandas DataFrame in Python. A first pass at the definition from Tristan is attached.
[TimeSeriesData_prototype.txt](https://github.com/BHoM/BHoM/files/5943539/TimeSeriesData_prototype.txt)
Also a useful reference: 
http://bluemountaincapital.github.io/Deedle/csharpseries.html#Time-series-sampling

**Comment by BingWangUS** [comment]
@rboulton-BH this is being prototyped in climate toolkit. A base Series oM is already created and some methods are being prototyped. A combination of method and time period is a interesting thought.

**Comment by FraserGreenroyd** [comment]
@enarhi to schedule a workshop on this for sprint 4 to discuss/flesh out more what is needed here functionality wise.

**Comment by tg359** [comment]
**Resurrecting this issue as I've been having a think about it lately, and it was the first one I found related to this workflow!**

## Where we are
We've now got `BH.oM.Environment.Result` objects which inherit from `BH.oM.Environment.IAnalysisResult` (`ITemperatureResult`, `IVelocityResult`, ...) and store a `double` which is then associated with the type of data being rpepresented by it's class. We also have a `BH.oM.Environments.TimeIndexResult` which can store `List<IAnalysisResult` along with a `TimeIndex` `int` for each of the `IAnalysisResult` objects being passed. 

## What's not currently possible
`TimeIndexResult` objects can be associated with `Node` objects, which can then be associated with meshes for visualisation (when that functionality becomes available) - however, I still have my concerns around how we associate time-indexed data with BHoM objects as the current set-up is very inefficient. For example, I created two lists of 8760 results and it took 1.6minutes just to generate the objects:
![image](https://user-images.githubusercontent.com/10939984/130482739-7c6800df-5e4d-4ccf-8ec5-2fdef0a42bd1.png) - which is more of an issue with how we're structuring the data in objects than what GH is doing. In reality, we'd have at least 1000  sets of annual hourly values which wouldn't be possible via the current set-up.

If we're working entirely in Rhino/GH, then this means a large number of objects being created just for the user to pass onto another stage in the process. We'd also likely be dealing with data generated elsewhere, so single objects containing the more complex value-time relationship would be better suited.

## What needs to be possible
We need to be able to run calculation on data generated by other sources (external simulations, results files generated by other tools, results sets stored in accessible DBs, etc.) The few interfaces we'd need for the initial AnalysisGrids would be spatially located results sets from Radiance (and other lighting tools) and potentially from dynamic thermal sims like IES and EnergyPlus (where zone-centroids are the grid-nodes). So we'd need interfaces for:

- .ill files (from Radiance, containing point-values results)
- .res files (from Radiance, containing point-value results)
- SQLite files (from EnergyPlus, containing point-value results)
- CSV (from anywhere, but mainly IES, containing point-values results)
- ... each with their own separate files storing the locations where those points exist

Once we can handle multidimensional data efficiently in BHoM objects without maxing out the RAM by filling it with hundreds of thousands of objects, we could then think about doing basic maths with those objects and passing them to render as meshes in whatever interface we want.

I'll link to this issue from where I find other related issues.

**Comment by tg359** [comment]
It's also worth noting that, as we're using Mongo, we should probably align our results objects so that they convert into something easily queryable in Mongo using the v5.0 time-series data capabilities here - https://docs.mongodb.com/manual/core/timeseries-collections/

**Comment by al-fisher** [comment]
This sounds really good @tg359 
Linking to https://github.com/BHoM/admin/issues/17 also as you plan the above out - will be good to map in the context of wider parent interfaces.

Agree with your comments above. 
In reviewing namespaces recently with @alelom as part of [above #17](https://github.com/BHoM/admin/issues/17) thought similar, that the Result objects in `BH.oM.Environment` have a lot of boiler plate for a single double.

One thing to already look at is removing the inheritance of `BHoMObject` from the results and implement the `IObject` interface instead.

`public class DryBulbTemperatureResult :` ~BHoMObject,~ `ITemperatureResult, IObject`

This is the pattern used in other discipline namespaces such as `BH.oM.Structure` - specifically to avoid all the additional properties `IBHoMObject` imposes. Which you really do not what when dealing with potentially 1000's and 1000's of entities as you say.

Both `IGeometry` and `IResult` interfaces implement only `IObject` for this reason.

@alelom FYI

**Comment by enarhi** [comment]
@tg359 thanks for resurrecting as this is critical to a lot of ongoing initiatives across a lot of disciplines it seems. Think what @al-fisher mentions here makes sense in terms of streamlining those objects, and I also totally agree with what you've flagged in terms of Mongo capability. 

One thing worth noting on that example you posted where it's taking 1.6 minutes: I suspect if you look at your results you have something akin to:
![image](https://user-images.githubusercontent.com/6752345/130506665-76dc694d-9b88-4002-b926-d9956017e7b8.png)

Looks like we have errant behavior with the TimeIndexResult due to it expecting a List of results per single time index value based on the object's properties, which will require a new create method that accounts for this (ie input list of list of results and single list of time indices, check list lengths match indices count, then create each result with the corresponding list of results per time indice provided). This would lead to performance as below, which is still not great given the required scaling up you mentioned but certainly isn't as impossible to work with!
![image](https://user-images.githubusercontent.com/6752345/130507508-f48f9c3c-8a7c-425c-8f2e-e538d1b6ecae.png)

**Comment by al-fisher** [comment]
In the [same admin issue linked to above](https://github.com/BHoM/admin/issues/17) - do take a look at **Analytical_oM** listed tasks in particular - and especially item…

> Reformulate the IResult interface, including creating a compositional set of interfaces: IResult, IObjectResult, ICaseResult, ITimedResult, Structure.IResult

Good to get your thoughts and input on this task with respect to requirements of time-series data and other Environment results

**Comment by tg359** [comment]
@al-fisher I'm interested in seeing what the `ITimedResult` from your admin issue might be, as that could be better placed than a discipline specific one (if the value storage aligns with our needs).

@enarhi I made the change you suggested, and it got faster (stupid GH list handling!) - then I changed to inherit from `IObject` too and that made a slight improvement. But if we could store lists in objects rather than a set of `TimeIndexResult` objects as are currently being created that would give us the best performance. 

I ran some tests under the changes I made above, and for a project I recently worked on that had 200,000 sensors each with 8760 values, it would take abut 10-hours (memory permitting) to create the `TimeIndexResults` via BHoM - and that's before we even start doing the maths with that 200000*8760 matrix!.

**Comment by alelom** [comment]
> @al-fisher I'm interested in seeing what the ITimedResult from your admin issue might be, as that could be better placed than a discipline specific one (if the value storage aligns with our needs).

Yes definitely. That's exactly why I'm trying to centralize more generic interfaces like that. I actually came up with TimedResult in the general effort of having the highest number of simplest interfaces possible, as it sounded like something useful. I haven't yet formalized what type it should enforce. I think I was kind of heading for C#'s `DateTime`. How do you generally store time yourself?

**Comment by tg359** [comment]
@alelom, I tend to do most (if not all) of my time-indexed stuff in Python using pandas to have datetime-indexed numpy array. In .NET this would be more akin to a nD array, where one of the dimensions aligns with a datetime-index of the same size so that once the mathsy bits have finished we can do the time-based aggregations.

Ideally, in BHoM, we'd have datetime stored in a 1d array next to a multidimensional array where the first dimension is the same length and the 1d array.

**Comment by alelom** [comment]
> Ideally, in BHoM, we'd have datetime stored in a 1D array, next to a multidimensional array where the first dimension is the same length as that 1D array.

So:
```cs
List<DateTime> timestamps
List<List<object>> data // where len(outer list) = len(timestamps)
```
?
